### PR TITLE
Add template-based bond determination

### PR DESCRIPTION
**Pull request by swails**
This is the start of the implementation of #34 (!!), and is the first step in ultimately tackling #355 and #662 

Things to do:

- [x] Handle atoms and residues not in templates
- [x] Figure out how to stitch together polymeric residues
- [x] Process CONECT records in PDB files ~~and the equivalent in mmCIF files~~
- [x] Implement the distance-based bonding for atoms and residues not present in templates
- [x] Special-case disulfide bonds and sulfur linkers

EDIT: More to-dos

- [x] Automatic bond assignment for GRO files
- [x] Improve element matching for GRO files based on residue templates (!! nice!)

EDIT: Even more

- [x] Implement a pairlist to drastically improve efficiency of bond searching

### CODE REVIEW

File: parmed/modeller/standardtemplates.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+""" Standard residue templates for biomolecular residues """
+
+import os
+from parmed.amber.offlib import AmberOFFLibrary as _parser
```
Comment (hainm) [inline parmed/modeller/standardtemplates.py]: you used _parser to avoid dumping AmberOFFLibrary in this namespace?

Comment (swails) [inline parmed/modeller/standardtemplates.py]: Yea, more or less. :smiley:

**Comment by swails** [comment]
OK, this should be done and good to go.  I'm punting on the mmCIF CONECT stuff for now, since I don't know the spec currently and the bond-by-distance stuff takes care of most of what that would miss (like disulfides, sulfide-linking, and non-standard residue bonding patterns.

I'm good to merge this when it passes.

Closes #34

**Comment by swails** [comment]
Actually, I found a couple more things to do.  Once I get to those, then this is good to merge.

**Comment by swails** [comment]
Ack... I really need to implement a fast, numpy-vectorized, grid-based pairlist generator for the unassigned atoms whose bonds are assigned by distance.  Otherwise the bond assignment scales as O(N^2).

**Comment by ctk3b** [comment]
[This](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/core/topology.py#L1022) might help a little, if I correctly understand what you're trying to do right now, but there's certainly more room for improvement.

**Comment by swails** [comment]
I considered that method, but I was worried about the potential memory and time requirements when systems get large. I was thinking something like a very simple grid/voxel-based approach.

The approach I'm thinking of should scale as O(N) in both memory and time with a reasonably small time prefactor (since the pairlist can be restricted to atoms within 2.5 angstroms of each other).  It should also be fast enough to do in pure Python...  But we'll see.  It would seem a shame to have to start requiring C extensions just for this.

**Comment by swails** [comment]
It's slower than I'd like, but acceptable for now.  It'll make parsing huge PDB files with many unknown residue types (such as those for bilayers and carbohydrates) ~4x slower (this is for the 117K atom CHARMM segid overflow PDB test with >60K atoms not in standard templates).

The pairlist is only built for atoms whose bonds still need to be assigned after going through the templates, so this should perform very well in the majority of cases, and acceptable (for now) in pathological ones.

**Comment by swails** [comment]
This should be good to go after the tests pass. Whew!

**Pull request merged**